src/Features/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs

```diff
@@ -203,6 +203,46 @@ public void UpdateTrackedDocument(Uri uri, SourceText newSourceText)
         return miscDocument;
     }
 
+    /// <summary>
+    /// Returns an analyzer config document with the LSP tracked text forked from the appropriate workspace solution.
+    /// 
+    /// This is always called serially in the <see cref="RequestExecutionQueue"/> when creating the <see cref="RequestContext"/>.
+    /// </summary>
+    public async Task<TextDocument?> GetLspAnalyzerConfigDocumentAsync(TextDocumentIdentifier textDocumentIdentifier, CancellationToken cancellationToken)
+    {
+        var uri = textDocumentIdentifier.Uri;
+
+        // Get the LSP view of all the workspace solutions.
+        var lspSolutions = await GetLspSolutionsAsync(cancellationToken).ConfigureAwait(false);
+
+        // Find the matching document from the LSP solutions.
+        foreach (var (lspSolution, isForked) in lspSolutions)
+        {
+            var additionalDocuments = lspSolution.GetAnalyzerConfigDocuments(uri);
+            if (additionalDocuments.Any())
+            {
+                var document = additionalDocuments.FindAnalyzerConfigDocumentInProjectContext(textDocumentIdentifier);
+
+                // Record metadata on how we got this document.
+                var workspaceKind = document.Project.Solution.Workspace.Kind;
+                _requestTelemetryLogger.UpdateFindDocumentTelemetryData(success: true, workspaceKind);
+                _requestTelemetryLogger.UpdateUsedForkedSolutionCounter(isForked);
+                _logger.TraceInformation($"{document.FilePath} found in workspace {workspaceKind}");
+
+                return document;
+            }
+        }
+
+        // We didn't find the document in any workspace, record a telemetry notification that we did not find it.
+        var searchedWorkspaceKinds = string.Join(";", lspSolutions.SelectAsArray(lspSolution => lspSolution.Solution.Workspace.Kind));
+        _logger.TraceError($"Could not find '{uri}'.  Searched {searchedWorkspaceKinds}");
+        _requestTelemetryLogger.UpdateFindDocumentTelemetryData(success: false, workspaceKind: null);
+
+        // Add the document to our loose files workspace if its open.
+        var miscDocument = _trackedDocuments.ContainsKey(uri) ? _lspMiscellaneousFilesWorkspace?.AddMiscellaneousDocument(uri, _trackedDocuments[uri], _logger) : null;
+        return miscDocument;
```

Comment:
surprising how complex this is.  if we're doing the same for normal documents, can we somehow share all this logic?

Reply:
this should be shareable - after all a `Document` is a `TextDocument`.  We should be able to share everything except for where we get the documents from the solution (`lspSolution.GetAnalyzerConfigDocuments(uri);` vs `var documents = lspSolution.GetDocuments(uri);`).

Likely we should have a helper method that takes in the function to get the documents from the solution
e.g.
`GetLspDocuments<T>(TextDocumentIdentifier textDocumentIdentifier, Func<T, Solution> documentGetter, CancellationToken cancellationToken) where T : TextDocument`